### ISSUE TITLE
bug : post process pipeline mismatch data preperation when has pad_shape

### ISSUE THREAD

**Issue by anxu829** [open-mmlab/mmsegmentation#3320]
when we consider resize process and  size divisibility in data preperation , your pipeline seems likie 

original image -> resize image -> pad img 

which refering to following shape record in img metas : 

(org_shape) -> (img_shape) -> (pad_shape)

but when you doing infering / evaluation , your pipeline seem likie this 

logits_shape -> resize to img_shape -> crop pad_size -> resize to org_shape , **which is not correct order **

below is a summary of your pipeline , which is misalign with the data preprocess order : 

check the annotation i write in the  predict function of EncoderDecoder : 

encoder_decoder.py

```
class EncoderDecoder:

    def predict(self, ...):
          # do inference here , and seg logit would be reshape to img_shape in this process
         seg_logits = self.inference(inputs, batch_img_metas)
         # do crop when has pad_shape , which is not correct !
         return self.postprocess_result(seg_logits, data_samples)

    def inference(self,inputs,batch_img_metas):
        seg_logit = self.whole_inference(inputs, batch_img_metas)
        return seg_logits
    def whole_inference(self,inputs , batch_img_metas):
        seg_logits = self.encode_decode(inputs, batch_img_metas)
        return seg_logits
    def encode_decode(self ,inputs, batch_img_metas)
            x = self.extract_feat(inputs)
            seg_logits = self.decode_head.predict(x, batch_img_metas,
                                                  self.test_cfg)
            return seg_logits
    def post_process(self , ...):
           padding_left, padding_right, padding_top, padding_bottom =padding_size
           # i_seg_logits shape is 1, C, H, W after remove padding
           i_seg_logits = seg_logits[i:i + 1, :,
                                          padding_top:H - padding_bottom,
                                          padding_left:W - padding_right]
           i_seg_logits = resize(
                    i_seg_logits,
                    size=img_meta['ori_shape'],
                    mode='bilinear',
                    align_corners=self.align_corners,
                    warning=False).squeeze(0)

class BaseDecodeHead:
    def predict(self,inputs , batch_img_metas):
        seg_logits = self.forward(inputs)
        return self.predict_by_feat(seg_logits, batch_img_metas)
   def predict_by_feat(self, seg_logits: Tensor,
                        batch_img_metas: List[dict]) -> Tensor:
        seg_logits = resize(
            input=seg_logits,
            size=batch_img_metas[0]['img_shape'],
            mode='bilinear',
            align_corners=self.align_corners)
       return seg_logits
```

**Comment by hujh1994** [comment]
No.
In decoder_head.predict_by_feat, if there is `pad_shape`, seg_logits is resized to pad_shape.
The logic is right.

But when there is not `pad_shape`, that is no padding is done in preprocess, seg_logits should not be resized to `img_shape`. This operation should be done in postprocess according to the design.
